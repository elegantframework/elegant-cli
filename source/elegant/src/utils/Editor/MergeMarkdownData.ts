import DOMPurify from 'dompurify';
import { Document } from '@/types/Document';
import UpdateImagePath from './UpdateImagePath';

export default function MergeMarkdownData(data: Document) {
  const meta = Object.entries(
    (({ content, publishedAt, ...meta }) => meta)(data)
  )

  if (data.publishedAt) {
    meta.push(['publishedAt', data.publishedAt.toISOString()])
  }

  let merged = '---\n'

  Object.entries(meta).forEach(([_, value]) => {
    if (Array.isArray(value[1])) {
      merged += `${value[0]}: ${JSON.stringify(value[1])}\n`
    } else if (value[1] instanceof Object) {
      merged += `${value[0]}:\n`
      Object.entries(value[1]).forEach(([key, value]) => {
        merged += `  ${key}: '${DOMPurify.sanitize(value as string).replaceAll(
          "'",
          "''"
        )}'\n`
      })
    } else {
      merged += `${value[0]}: '${DOMPurify.sanitize(value[1]).replaceAll(
        "'",
        "''"
      )}'\n`
    }
  })

  merged += '---\n\n';

  merged += UpdateImagePath(data.content);

  return merged;
}