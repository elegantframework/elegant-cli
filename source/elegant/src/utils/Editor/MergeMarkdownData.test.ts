import MergeMarkdownData from './MergeMarkdownData';

describe('Merge Markdown Data method', () => {
    it('returns the file as expected',() => {

      let result = MergeMarkdownData({
        title: 'Welcome',
        status: 'published',
        author: {
            name: 'Brandon Owens',
            picture: 'https://avatars.githubusercontent.com/u/10189130?v=4'
        },
        slug: 'welcome',
        description: 'Welcome to Elegant.',
        coverImage: '/images/social-card-large-EyMT.jpg',
        content: '\
            \
            ## Welcome to Elegant!\
            \
            We are excited that you are here ðŸ˜Š\
            \
            \
            \
            <div class="my-8 shadow-xl"><img class="rounded-xl" src="/api/admin/images/screenrecorderproject28--1--AwNz.png" alt=""></div>\
        '
      });

      expect(result).toMatchSnapshot(); 
    });
});