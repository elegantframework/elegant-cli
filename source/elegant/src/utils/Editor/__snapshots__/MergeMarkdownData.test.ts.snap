// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge Markdown Data method returns the file as expected 1`] = `
"---
title: 'Welcome'
status: 'published'
author:
  name: 'Brandon Owens'
  picture: 'https://avatars.githubusercontent.com/u/10189130?v=4'
slug: 'welcome'
description: 'Welcome to Elegant.'
coverImage: '/images/social-card-large-EyMT.jpg'
---

                        ## Welcome to Elegant!                        We are excited that you are here ðŸ˜Š                                                <div class="my-8 shadow-xl"><img class="rounded-xl" src="/images/screenrecorderproject28--1--AwNz.png" alt=""></div>        "
`;
